src/ERC6538Registry.sol

```diff
@@ -1,21 +1,65 @@
-// SPDX-License-Identifier: UNLICENSED
-pragma solidity 0.8.20;
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.23;
 
 import {IERC6538Registry} from "./interfaces/IERC6538Registry.sol";
 
 /// @dev `ERC6538Registry` contract to map accounts to their stealth meta-address. See
 /// [ERC-6538](https://eips.ethereum.org/EIPS/eip-6538) to learn more.
 contract ERC6538Registry is IERC6538Registry {
-  /// @notice Maps a registrant's identifier to the scheme ID to the stealth meta-address.
+  /// @notice Next nonce expected from `user` to use when signing for `registerKeysOnBehalf`.
   /// @dev `registrant` may be a standard 160-bit address or any other identifier.
   /// @dev `schemeId` is an integer identifier for the stealth address scheme.
-  mapping(bytes registrant => mapping(uint256 schemeId => bytes)) public stealthMetaAddressOf;
+  mapping(address registrant => mapping(uint256 schemeId => bytes stealthMetaAddress)) public
+    stealthMetaAddressOf;
+
+  /// @notice A nonce used to ensure a signature can only be used once.
+  /// @dev `user` is the registrant address.
+  /// @dev `nonce` will be incremented after each valid `registerKeysOnBehalf` call.
+  mapping(address user => uint256 nonce) public nonceOf;
+
+  /// @dev EIP-712 Type hash used in `registerKeysOnBehalf`.
+  bytes32 public constant TYPE_HASH =
+    keccak256("EIP712Domain(string name,string version,uint256 chainId,address registryContract)");
```

Comment:
Wait, isn't the typehash supposed to reflect the data that's being signed over?

Reply:
I think this is correct, it's an input to computing domain separator. I like using solmate as a reference implementation, it's much more readable than OZ (and to your point elsewhere, we can copy out that code instead of using OZ's: https://github.com/transmissions11/solmate/blob/0384dbaaa4fcb5715738a9254a7c0a4cb62cf458/src/tokens/ERC20.sol#L170)

Reply:
Implemented solmate style EIP712 signature verification in [a5898bd](https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5/commits/a5898bd3ed0e30960e0f17e686d19a073b0ddb62). There's a slight difference in how I've implemented tho, particularly how we set the `dataHash` and `recoveredAddress` so we can use it down a few more lines outside the bounds of `unchecked`.

Reply:
Ah whoops, yes you are correct. Here's the example from the [EIP itself](https://eips.ethereum.org/EIPS/eip-712#rationale-for-typehash)
<img width="738" alt="image" src="https://github.com/ScopeLift/stealth-address-erc-contracts/assets/17163988/1eb0a907-36b4-40b8-91fe-c7637c16ab2b">

Reply:
Looks like I mixed up two different types of typeHash. Should be fixed in [d8140ae](https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5/commits/d8140ae1146715c9465d85b2a86e534441e36f75)

Reply:
OK! This looks right now. We have the domain separator and the type hash and we're using them appropriately! Thanks for getting that cleaned up. A few thoughts though:

* I don't think we want/need to include the `registrant` (aka the signer) in the type hash data. Having the signer sign over itself seems redundant unless I'm missing something?
* I think the name of the type hash structure should be a noun, right? You've set it to match the method name but the examples I see don't follow that convention. I had suggested `Erc6538RegistryEntry` but open to something else.
* You moved the definition of the type hash into the method itself as a literal. I was wondering why you did that? I think it might help would-be integrators to have a public constant with the type hash available, unless you see some downside?

Reply:
> * I don't think we want/need to include the `registrant` (aka the signer) in the type hash data. Having the signer sign over itself seems redundant unless I'm missing something?

You're right, it is redundant. I've removed it.

> * I think the name of the type hash structure should be a noun, right? You've set it to match the method name but the examples I see don't follow that convention. I had suggested `Erc6538RegistryEntry` but open to something else.

Looking at the examples, noun seems to be the convention. Renamed it to `Erc6538RegistryEntry`.


> * You moved the definition of the type hash into the method itself as a literal. I was wondering why you did that? I think it might help would-be integrators to have a public constant with the type hash available, unless you see some downside?

Initially the definition was created because it was used in two places, but I then moved them into the method after realizing they are two different hash and each only occur once. But I understand we have to think of the future integrators, so I've added back the type hash definition as:
```js
/// @notice The EIP-712 type hash used in `registerKeysOnBehalf`.
bytes32 public constant ERC6538REGISTRY_ENTRY_TYPE_HASH =
keccak256("Erc6538RegistryEntry(uint256 schemeId,bytes stealthMetaAddress,uint256 nonce)");
```